doc/source/dev/roadmap-detailed.rst

```diff
@@ -333,27 +333,44 @@ The sparse matrix formats are mostly feature-complete, however the main issue
 is that they act like ``numpy.matrix`` (which will be deprecated in NumPy at
 some point).
 
-What we want is sparse arrays, that act like ``numpy.ndarray``. In SciPy
-``1.8.0`` a new set of classes (``csr_array`` et al.) has been added - these
-need testing in the real world, as well as a few extra features like 1-D array
-support.
-An alternative (more ambitious, and unclear if it will materialize at this
-point) plan is being worked on in https://github.com/pydata/sparse.  The
-tentative plan for that was/is:
-
-- Start depending on ``pydata/sparse`` once it's feature-complete enough (it
-  still needs a CSC/CSR equivalent) and okay performance-wise.
-- Add support for ``pydata/sparse`` to ``scipy.sparse.linalg`` (and perhaps to
-  ``scipy.sparse.csgraph`` after that).
-- Indicate in the documentation that for new code users should prefer
-  ``pydata/sparse`` over sparse matrices.
-- When NumPy deprecates ``numpy.matrix``, vendor that or maintain it as a
-  stand-alone package.
+What we want is sparse arrays that act like ``numpy.ndarray``. Initial
+support for a new set of classes (``csr_array`` et al.) was added in SciPy
+``1.8.0`` and stabilized in ``1.12.0`` when construction functions for
+arrays were added. Support for 1-D array is expected in ``1.13.0``.
+
+Next steps toward sparse array support:
+
+- Extend sparse array API to 1-D arrays.
+    - Support for COO, CSR and DOK formats.
+    - CSR 1D support for min-max, indexing, arithmetic.
```

Comment:
I'm still not sure I understand the value of a public 1d sparse CSR. I see how it's useful to re-use the CSR C++ code internally for 1d COO, but I see this more as an implementation detail. Is there further discussion of this somewhere?

Reply:
Having separate classes for each format means we have different methods (features as well as implementation details). Each format has different features (not implementation details). For example COO doesn't support indexing/minmax/binop. The most feature-complete formats are CSR and CSC. So having the CSR format support 1d arrays makes a lot of sense from the feature-set perspective.

We looked hard at having a separate format for 1d arrays (call it UNI for universal, or unified or unidimensional or whatever). The UNI format has the feature-set of CSR (that is, of `csr_array`) with the storage structure of COO. But the "implementation details" get in the way -- perhaps there is a better way to implement it. It's built as a separate class that subclasses coo_array and csr_array. The public API calls `coo_array(arg)` and chooses whether to instantiate a uni_array or a coo_array based on the dimension requested. But it is hard to shift between shapes and sizes if the class of the object needs to change. For example, `A.resize()` has to change the dimension in-place! I don't think it's a good idea to change the class of an object in-place, if it is even possible. That was a hurdle that made me stop working on uni_array. But there were going to be other similar issues. 

From a bigger perspective, the dimension and shape of an array are "details" that don't seem to require a change in format for the libraries I have worked with. Just as CSR is generalized to nd-arrays in some libraries, CSR is generalized to 1d-arrays too. If it helps, think of the list of compressed axes and uncompressed axes -- the 1d case has `[]` for the compressed axes and `[0]` for the uncompressed axes. 

Implementation details matter. They impact api-design. In this case they impact it by having different formats/classes support different feature sets. The 1d case can fit into that structure -- and does so nicely.

Reply:
Just for a quick clarification:

> Each format has different features (not implementation details). For example COO doesn't support indexing/minmax/binop.

Is this fundamental? Clearly one could do these things with COO (as many libraries do), scipy just doesn't have these features implemented.

If COO had these features implemented, would you still want 1d CSR?

Reply:
I can only say what I think it should be -- this isn't necessarily what others would choose.

I don't think all features should be available for all formats. That's the whole reason for the different storage formats. Some are just better for doing certain operations. For example, I don't think matrix multiplication should be allowed from the COO format. An error should be raised telling the user to convert to CSR or CSC before multiplying. In a real sense, allowing the user to do e.g. matrix multiplication from a coo_array is doing the user a disservice. They will end up doing many automatic unnecessary conversions rather than converting once and doing multiple operations. But if we help them learn which formats do what operations well -- and restrict which can be done by which classes via helpful messages -- they will build workflows that convert when converting is effective.

Another approach would be to just have one public format. The storage gets converted or cached automatically under the hood. If all formats do all operations (and automatically convert when more efficient) then we would make one `sparse_array` public class. That is certainly nice if it ends up being performant. But I'm not aware of good algorithms for knowing when to convert automatically.

So, I guess I am saying that "this is fundamental". The different storage formats are better for different things. Given my current experience with these issues, I would choose that storage formats are aligned with feature sets that work for them.  And various `ndim`s are supported where it is performant to do so.

But you probably have more experience and a different perspective. What would you want to be different between the various format classes? Should we aim for all features for all classes?

Reply:
I think as much of the array-api as makes sense should be available for `coo`, `csr`, and `csc`.

>  I don't think matrix multiplication should be allowed from the COO format. An error should be raised telling the user to convert to CSR or CSC before multiplying. In a real sense, allowing the user to do e.g. matrix multiplication from a coo_array is doing the user a disservice. They will end up doing many automatic unnecessary conversions rather than converting once and doing multiple operations.

A few points here

* COO can be faster here, especially for very sparse operations. I think throwing a efficiency warning if we can detect it's fairly dense could be reasonable.
* Currently the library does internal conversions between classes quite frequently. A few examples:
    * Conversion is the default for most methods under `_spbase`
    * `min`/ `max` converting to csc/ csr based on axis being reduced along https://github.com/scipy/scipy/blob/fb4c65fdf89866728a9762b47e1c75c03e0b2777/scipy/sparse/_data.py#L186
    * CSC/ CSR internally converting to COO in `.multiply`: https://github.com/scipy/scipy/blob/fb4c65fdf89866728a9762b47e1c75c03e0b2777/scipy/sparse/_compressed.py#L438

> Another approach would be to just have one public format. The storage gets converted or cached automatically under the hood. If all formats do all operations (and automatically convert when more efficient) then we would make one `sparse_array` public class. That is certainly nice if it ends up being performant. But I'm not aware of good algorithms for knowing when to convert automatically.

I think this need a compiler and to know ahead which operations to do. Also seems like it could be memory intensive. The nice thing about the current api is that it's easy to extend with custom code since the format is explicit.

> So, I guess I am saying that "this is fundamental". The different storage formats are better for different things. Given my current experience with these issues, I would choose that storage formats are aligned with feature sets that work for them. And various `ndim`s are supported where it is performant to do so.

I do agree storage formats are good for different things. But I do think COO is the right choice for performance in a number of cases, especially `nd` and anything hypersparse.

I think it would also make sense to do some limiting of features for different arrays, adding some extra features for some. But I think most features from the array-api should work for `coo`, `csr`, and `csc`. Notably, I think many indexing like features can't work for `csr`, `csc` because they are 2d.

I wouldn't think adding support for every method is worth it for classes like DOK.

---

One other performace case to mention:

`coo.tocsr()` makes a copy of the `indices` and `data`. For an operation like `dense += coo` if the user had to convert to `csr` that could be a significant allocation that is unnecessary for the task.

Reply:
Just to see if I understand what you are saying:
- converting to other formats (especially without copying arrays wherever it is not needed) can be helpful in many situations and is sometimes automatically done in SciPy currently.
- automatic conversion is not always best -- so it is nice for users to determine whether conversion occurs or not -- with a good default. (That is, assuming that the current format supports the operation.)
- it's hard to write automated decision tools for when to convert.

I agree with all of these thoughts. Your push for more features in COO is clear. But some of your justifications for COO seems to limit what CSR and CSC can do. There is no reason to restrict compressed storage to 2d. You can hold a tuple of axes that are compressed and a tuple that are not compressed. (I guess there must be at least one axis that is not compressed, but I haven't thought about that).  Ravel the indices from the compressed axes to give the compressed info (call it `indptr` for example). Do the same for the uncompressed axes and you've got `indices`. Now you have some machinery to ravel and unravel indices, and you have some machinery to do operations on the raveled versions of the indices. In this setting, CSR, CSC and COO are three possible choices for compressed and uncompressed axes. The COO doesn't have compressed axes. The other two have 1 of each axis type. Then 1d, 3d, 4d, etc can all be described in this framework of compressed and uncompressed axes. It seems reasonable to even cache the axes that need to convert back and forth often. So those axes have two versions of the indices stored.  This is the spirit in which the Scipy csr_array supports 1d arrays.  And pydata-sparse uses this spirit for their nD compressed arrays. Restricting CSR and CSC to 2d is technically correct, but the spirit of the format extends to all shapes.

COO with sorted/deduplicated indices make changes to the array slow. I think that is the main justification for DOK and dod formats. The dicts allows us to make changes quickly. Then once the array is constructed we convert to csr for the operations. coo can play this role too. I haven't compared dok and coo very much. 

From the network/graph perspective, COO is an edgelist data structure. CSR is a adjacency data structure. Many graph algorithms work well on adjacency data structures, but there are some where edgelist is better. But edgelists are easier to think about and manipulate. So, my impression with network analysis is that you wantto communicate with humans using edge lists, but usually compute things with adjacencies. Swapping that over to sparse my instinct is to use COO for communication and manipulation while using csr for algorithms. But I don't have enough experience to know how far that can be pushed.

Your example of `coo.tocsr()` a suggests a performance improvement. The C++ code is set up to copy indices and data (perhaps because the order of the values may change as you expand the compressed axis indices). If there is a way to decompress the axis without changing the order of the indices and data info, we should do that instead of the current code. Do you know how to order things correctly for that? Perhaps the COO needs to be sorted in a specific way.